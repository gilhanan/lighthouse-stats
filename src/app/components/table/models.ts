export interface Cell {
  label: string;
  value: unknown | Cell[];
}

export interface Row {
  id: string;
  cells: Cell[];
}
