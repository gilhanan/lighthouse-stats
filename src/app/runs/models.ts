export interface HostParams {
  host: string;
}

export interface ProjectParams extends HostParams {
  project: string;
}

export interface BranchParams extends ProjectParams {
  branch: string;
}

export interface BuildParams extends ProjectParams {
  build: string;
}

export interface RunsParams extends BuildParams {
  url: string;
}

export interface BuildFormState {
  host?: string;
  project?: Project;
  url?: URL;
  branch?: Branch;
  build?: Build;
}

export interface Project {
  id: string;
  name: string;
  baseBranch: string;
}

export interface Branch {
  branch: string;
}

export interface URL {
  url: string;
}

export interface Build {
  id: string;
  branch: string;
  commitMessage: string;
}

export interface RunsFormState {
  category?: Category;
}

export interface RawRun {
  id: string;
  projectId: string;
  buildId: string;
  representative: boolean;
  url: string;
  lhr: string;
}

interface Audit {
  id: string;
  title: string;
  score: number;
  numericValue: number;
}

export interface Category {
  id: string;
  title: string;
  score: number;
  auditRefs: { id: string; weight: string }[];
}

export interface LHR {
  environment: {
    benchmarkIndex: number;
  };
  audits: Record<string, Audit>;
  categories: Record<string, Category>;
}

export type Run = RawRun & {
  lhr: LHR;
};
