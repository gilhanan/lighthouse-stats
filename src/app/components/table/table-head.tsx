import React from "react";
import { Cell } from "./models";

function calculateDepth(cells: Cell[]): number {
  return (
    1 +
    Math.max(
      0,
      ...cells.map(({ value }) =>
        Array.isArray(value) ? calculateDepth(value) : 0
      )
    )
  );
}

function getCellsAtDepth({
  cells,
  depth,
  currentDepth = 1,
}: {
  cells: Cell[];
  depth: number;
  currentDepth?: number;
}): Cell[] {
  const result: Cell[] = [];

  if (depth < 1) {
    return result;
  }

  cells.forEach((cell) => {
    if (currentDepth === depth) {
      result.push(cell);
    } else if (Array.isArray(cell.value)) {
      result.push(
        ...getCellsAtDepth({
          cells: cell.value as Cell[],
          depth,
          currentDepth: currentDepth + 1,
        })
      );
    }
  });

  return result;
}

function getColSpan({ value }: Cell): number {
  if (!Array.isArray(value)) {
    return 1;
  }

  return (value as Cell[]).reduce(
    (sum, childCell) => sum + getColSpan(childCell),
    0
  );
}

export function TableHead({ cells }: { cells: Cell[] }) {
  const maxDepth = calculateDepth(cells);

  const rows = new Array(maxDepth).fill(null).map((_, rowIndex) => {
    const currentCells = getCellsAtDepth({ cells, depth: rowIndex + 1 });

    return (
      <tr key={rowIndex}>
        <th></th>
        {currentCells.map((cell, cellIndex) => {
          const { label, value } = cell;
          const rowSpan = Array.isArray(value) ? 1 : maxDepth - rowIndex;
          const colSpan = getColSpan(cell);

          return (
            <th
              key={cellIndex}
              rowSpan={rowSpan}
              colSpan={colSpan}
              className="text-center border p-2 bg-gray-50"
            >
              {label}
            </th>
          );
        })}
      </tr>
    );
  });

  return <thead>{rows}</thead>;
}
