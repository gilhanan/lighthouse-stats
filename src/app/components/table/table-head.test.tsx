import { render } from "@testing-library/react";
import { TableHead } from "./table-head";
import { Cell } from "./models";

const cells: Cell[] = [
  {
    label: "Simple",
    value: 0,
  },
  {
    label: "Two animals",
    value: [
      {
        label: "Cat",
        value: 1,
      },
      {
        label: "Dog",
        value: 2,
      },
    ],
  },
  {
    label: "Three vegetables",
    value: [
      {
        label: "Tomato",
        value: 3,
      },
      {
        label: "Potato",
        value: 4,
      },
      {
        label: "Carrot",
        value: 5,
      },
    ],
  },
  {
    label: "Three levels",
    value: [
      {
        label: "First",
        value: [
          {
            label: "One",
            value: 6,
          },
          {
            label: "Two",
            value: 7,
          },
        ],
      },
      {
        label: "Second",
        value: [
          {
            label: "Three",
            value: 8,
          },
          {
            label: "Four",
            value: 9,
          },
        ],
      },
    ],
  },
];

it("renders homepage unchanged", () => {
  const { container } = render(<TableHead cells={cells} />);
  expect(container).toMatchSnapshot();
});
