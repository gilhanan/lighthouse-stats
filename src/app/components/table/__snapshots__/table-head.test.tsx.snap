// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <thead>
    <tr>
      <th />
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="3"
      >
        Simple
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="2"
        rowspan="1"
      >
        Two animals
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="3"
        rowspan="1"
      >
        Three vegetables
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="4"
        rowspan="1"
      >
        Three levels
      </th>
    </tr>
    <tr>
      <th />
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="2"
      >
        Cat
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="2"
      >
        Dog
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="2"
      >
        Tomato
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="2"
      >
        Potato
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="2"
      >
        Carrot
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="2"
        rowspan="1"
      >
        First
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="2"
        rowspan="1"
      >
        Second
      </th>
    </tr>
    <tr>
      <th />
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="1"
      >
        One
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="1"
      >
        Two
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="1"
      >
        Three
      </th>
      <th
        class="text-center border p-2 bg-gray-50"
        colspan="1"
        rowspan="1"
      >
        Four
      </th>
    </tr>
  </thead>
</div>
`;
